include/permanent.hpp

```diff
@@ -322,34 +322,253 @@ std::complex<double> permanent_quad(std::vector<std::complex<double>> &mat) {
+    llint x = static_cast<llint>(pow(2,n - 1));
+    std::vector<T> col_sum(n, static_cast<T>(0));
+
+    // init col_sum 
+    for (int i=0; i < n; ++i) {
+        for (int j=0; j < n; ++j) {
+            col_sum[i] += mat[j*n + i];
+        }
+    }
+
+    constexpr T p1 = static_cast<T>(1.0);
+    constexpr T p2 = static_cast<T>(2.0);
+    constexpr T n2 = static_cast<T>(-2.0);
+
+    T mulcolsum, coeff;
+    T total = static_cast<T>(0);
+    llint og=0, ng, gd;
+    int sgn=1, gdi;
+
+    // over all 2^{n-1} permutations of delta
+    for (llint k=1; k < x+1; ++k) {
+        mulcolsum = std::accumulate(
+                        col_sum.begin(), 
+                        col_sum.end(), 
+                        p1, 
+                        std::multiplies<T>());
+        total += sgn > 0 ? mulcolsum : -mulcolsum;
+        
+        // updating gray order
+        ng = igray(k);
+        gd = og ^ ng;
+
+        coeff = og > ng ? p2 : n2;
+        gdi = left_most_set_bit(gd);
+        for (int j=0; j < n; ++j) {
+            col_sum[j] += coeff * mat[gdi * n + j];
+        }
+
+        sgn = -sgn;
+        og = ng;
+    }
+
+    // TODO: overflow handling
+
+    return total / static_cast<T>(x);
+}
+
+/**
+ * Returns the permanent of a matrix (nthreads=1) (2nd version)
+ *
+ * \rst
+ *
+ * Returns the permanent of a matrix using the BBFG algorithm.
+ * This algorithm was given by Glynn (2010) with the time-complexity 
+ * of \f$O(n 2^n)\f$ using Gray code ordering.
+ *
+ * \endrst
+ *
+ *
+ * @param mat  a flattened vector of size \f$n^2\f$, representing an
+ *      \f$n\times n\f$ row-ordered symmetric matrix.
+ * @return permanent of the input matrix 
+ */
+template <typename T>
+inline T perm_BBFG_serial2(std::vector<T> &mat)
+{
+    int n = std::sqrt(static_cast<double>(mat.size()));
+    int sgn=1, k=0;
+
+    constexpr T p2 = static_cast<T>(2.0);
+    constexpr T p1 = static_cast<T>(1.0);
+
+    int *gray_list = new int[n];
+    T *coeffs = new T[n];
+    std::fill_n(coeffs, n, p2);
+
+    std::vector<T> col_sum(n, static_cast<T>(0));
+    T mulcolsum, total = p1; 
+
+    // init col_sum 
+    for (int i=0; i < n; ++i) {
+        gray_list[i] = i;
+        for (int j=0; j < n; ++j) {
+            col_sum[i] += mat[j*n + i];
+        }
+        total *= col_sum[i];
+    }
+
+    while (k < n-1) {
+        for (int j=0; j < n; ++j) {
+            col_sum[j] -= coeffs[k] * mat[k*n+j];
+        }
+        
+        mulcolsum = std::accumulate(
+                        col_sum.begin(), 
+                        col_sum.end(), 
+                        p1, 
+                        std::multiplies<T>());
+
+        coeffs[k] = -coeffs[k];
+        sgn = -sgn; 
+        total += sgn > 0 ? mulcolsum : -mulcolsum;
+
+        // update ordering 
+        gray_list[0] = 0;
+        gray_list[k] = gray_list[k+1];
+        gray_list[k+1] = k+1;
+        k = gray_list[0];
+    }
+
+    delete[] gray_list;
+    delete[] coeffs;
```

Comment:
If an exception is thrown during the above method that is caught by the bound Python code, are these lines ever hit?

Reply:
If an exception is thrown after the allocation and before freeing, this code definitely causes memory-leaks! I used the  pair of new-delete in the `*_serial2` function only to test my theory of how using lists to perform perm. ordering and column-sum coefficients could increase the performance of the BBFG permanent function. I just fixed this issue.